### DISCUSSION TITLE
Improve poetry versioning re: pep440/npm

### DISCUSSION THREAD

**Discussion by rarkins** [renovatebot/renovate#24802] [Suggest an Idea]
**What would you like Renovate to be able to do?**

Coverage a wider set of poetry ranges/versions than it does currently.

**Did you already have any implementation ideas?**

First, stop deciding between pep440 and poetry versioning during poetry extract and instead always assign poetry.

Then, poetry versioning should auto-detect whether to call npm or pep440, using per-repo in-mem caching so that it's not too inefficient.

We should enforce `pep440.isVersion()` and not use npm. If someone has a dependency with strange versioning then they will need to change versioning manually using packageRules.

Next, when we encounter a range we should:
- Check if we've seen it before in this repo and use the cached result if so
- If not, check if it's already a valid pep440 range and note that if so. Pass it unchanged to pep440 functions
- If not, check the result of `npm.isValid(poetry2npm(range))`. It should return something like `>=2.0.0 <3.0.0` if it's valid. If *that* string is a valid pep440 range then use it, but otherwise we massage it into pep440 syntax and use that instead (if it validates). Save the result!

The end result is that the only time we use `npm` versioning is for massaging poetry's npm-like syntax into pep440 range syntax.

**Comment by viceice** [comment]
As poetry uses semver v2 for normal deps, we should do so too and only fall back to pep440 is it's not valid semver?

**Comment by rarkins** [comment]
It don't think that's really the case because ultimately the versions are sourced from pypi and it doesn't enforce semver versions. I've seen dozens or maybe hundreds of cases so far where poetry is used to install a dependency which has versions valid with pep440 but not semver.

We could maybe study the poetry source to be sure or ask the maintainers for advice if we can't tell.

**Comment by MindTooth** [comment]
I'm sorry if this is not related.  I see that recently you've been active regarding poetry and felt maybe I could post it here.

Adding black[1] using `poetry add black` results it using a semver range (`^20.8b1`) with a PEP440 version.  It does not seem that Renovate is able to parse this.  I tried various combinations of matchers using the docs and some examples here on GitHub as inspiration but to no avail.

---

```JSON
{
  "depName": "black",
  "depType": "dev-dependencies",
  "currentValue": "^20.8b1",
  "managerData": { "nestedVersion": false },
  "datasource": "pypi",
  "skipReason": "unknown-version",
  "depIndex": 4,
  "updates": []
}
```

[1]: https://black.readthedocs.io/en/stable/
